File: Autonomous_Mobile_Robots_Course/Chapter7/Lesson1/Chapter7_Lesson1.m
% Chapter7_Lesson1.m
% Chapter 7 - Lesson 1: EKF Localization Pipeline (mobile framing)
%
% Minimal EKF localization demo (2D) in MATLAB (from scratch).
% State: [x; y; theta] in world frame, Control: [v; omega] in body frame
% Measurement: range-bearing to known landmarks.

function Chapter7_Lesson1()
    rng(7);

    landmarks = [ 5 0;
                  5 5;
                  0 5;
                 -3 2 ];

    dt = 0.1;
    N  = 250;

    % True noise on control (simulation)
    sigma_v_true = 0.05;
    sigma_w_true = 0.03;

    % Assumed process noise in EKF (tuning)
    Q = diag([0.08^2, 0.05^2]);

    % Measurement noise (range, bearing)
    R = diag([0.15^2, (2*pi/180)^2]);

    % Initial truth and estimate
    x_true = [0; 0; 0];
    x_est  = [-0.2; 0.1; 0.05];
    P      = diag([0.5^2, 0.5^2, (10*pi/180)^2]);

    sumPos2 = 0;
    sumTh2  = 0;

    for k = 1:N
        v_cmd = 0.8 + 0.2*sin(0.04*(k-1));
        w_cmd = 0.25 + 0.05*cos(0.03*(k-1));
        u_cmd = [v_cmd; w_cmd];

        u_true = [v_cmd + sigma_v_true*randn();
                  w_cmd + sigma_w_true*randn()];

        % True propagation
        x_true = motion_model(x_true, u_true, dt);

        % EKF predict
        F = jacobian_F(x_est, u_cmd, dt);
        L = jacobian_L(x_est, dt);
        x_est = motion_model(x_est, u_cmd, dt);
        P = F*P*F' + L*Q*L';
        P = 0.5*(P + P');

        % Measurement (nearest landmark)
        idx = nearest_landmark(x_true, landmarks);
        lm  = landmarks(idx,:);

        z_true = meas_model(x_true, lm);
        z_meas = [z_true(1) + sqrt(R(1,1))*randn();
                  wrap_angle(z_true(2) + sqrt(R(2,2))*randn())];

        % EKF update
        H = jacobian_H(x_est, lm);
        zhat = meas_model(x_est, lm);

        y = z_meas - zhat;
        y(2) = wrap_angle(y(2));

        S = H*P*H' + R;
        K = P*H'/S;

        x_est = x_est + K*y;
        x_est(3) = wrap_angle(x_est(3));

        I = eye(3);
        P = (I - K*H)*P*(I - K*H)' + K*R*K';
        P = 0.5*(P + P');

        ex = x_est(1) - x_true(1);
        ey = x_est(2) - x_true(2);
        eth = wrap_angle(x_est(3) - x_true(3));
        sumPos2 = sumPos2 + ex^2 + ey^2;
        sumTh2  = sumTh2 + eth^2;
    end

    rmsPos = sqrt(sumPos2 / N);
    rmsTh  = sqrt(sumTh2  / N);

    fprintf('RMS position error [m]: %.4f\n', rmsPos);
    fprintf('RMS heading error [rad]: %.4f\n', rmsTh);
end

function a = wrap_angle(a)
    a = mod(a + pi, 2*pi) - pi;
end

function xnext = motion_model(x, u, dt)
    px = x(1); py = x(2); th = x(3);
    v = u(1);  w  = u(2);
    xnext = zeros(3,1);
    xnext(1) = px + dt*v*cos(th);
    xnext(2) = py + dt*v*sin(th);
    xnext(3) = wrap_angle(th + dt*w);
end

function F = jacobian_F(x, u, dt)
    th = x(3);
    v  = u(1);
    F = eye(3);
    F(1,3) = -dt*v*sin(th);
    F(2,3) =  dt*v*cos(th);
end

function L = jacobian_L(x, dt)
    th = x(3);
    L = zeros(3,2);
    L(1,1) = dt*cos(th);
    L(2,1) = dt*sin(th);
    L(3,2) = dt;
end

function z = meas_model(x, lm)
    px = x(1); py = x(2); th = x(3);
    mx = lm(1); my = lm(2);
    dx = mx - px;
    dy = my - py;
    r  = sqrt(dx^2 + dy^2);
    b  = wrap_angle(atan2(dy, dx) - th);
    z = [r; b];
end

function H = jacobian_H(x, lm)
    px = x(1); py = x(2);
    mx = lm(1); my = lm(2);
    dx = mx - px;
    dy = my - py;
    q  = dx^2 + dy^2;
    eps = 1e-12;
    if q < eps, q = eps; end
    r  = sqrt(q);

    H = zeros(2,3);
    % range
    H(1,1) = -dx/r;
    H(1,2) = -dy/r;
    % bearing
    H(2,1) =  dy/q;
    H(2,2) = -dx/q;
    H(2,3) = -1;
end

function idx = nearest_landmark(x, landmarks)
    dx = landmarks(:,1) - x(1);
    dy = landmarks(:,2) - x(2);
    d  = sqrt(dx.^2 + dy.^2);
    [~, idx] = min(d);
end

File: Autonomous Mobile Robots/Chapter7/Lesson1/Chapter7_Lesson1.java
// Chapter7_Lesson1.java
// Chapter 7 - Lesson 1: EKF Localization Pipeline (mobile framing)
//
// Minimal EKF localization demo (2D) in Java using EJML.
// Dependency: org.ejml:ejml-simple
//
// Compile/run (example with Gradle or Maven):
// - Add EJML dependency and run this class' main().

import org.ejml.simple.SimpleMatrix;

import java.util.Random;

public class Chapter7_Lesson1 {

    static class Landmark {
        final double mx, my;
        Landmark(double mx, double my) { this.mx = mx; this.my = my; }
    }

    static double wrapAngle(double a) {
        double twoPi = 2.0 * Math.PI;
        a = (a + Math.PI) % twoPi;
        if (a < 0) a += twoPi;
        return a - Math.PI;
    }

    static SimpleMatrix motionModel(SimpleMatrix x, SimpleMatrix u, double dt) {
        double px = x.get(0), py = x.get(1), th = x.get(2);
        double v = u.get(0), w = u.get(1);
        SimpleMatrix nx = new SimpleMatrix(3,1);
        nx.set(0, px + dt * v * Math.cos(th));
        nx.set(1, py + dt * v * Math.sin(th));
        nx.set(2, wrapAngle(th + dt * w));
        return nx;
    }

    static SimpleMatrix jacobianF(SimpleMatrix x, SimpleMatrix u, double dt) {
        double th = x.get(2);
        double v  = u.get(0);
        SimpleMatrix F = SimpleMatrix.identity(3);
        F.set(0,2, -dt * v * Math.sin(th));
        F.set(1,2,  dt * v * Math.cos(th));
        return F;
    }

    static SimpleMatrix jacobianL(SimpleMatrix x, double dt) {
        double th = x.get(2);
        SimpleMatrix L = new SimpleMatrix(3,2);
        L.set(0,0, dt * Math.cos(th));
        L.set(1,0, dt * Math.sin(th));
        L.set(2,1, dt);
        return L;
    }

    static SimpleMatrix measRangeBearing(SimpleMatrix x, Landmark lm) {
        double dx = lm.mx - x.get(0);
        double dy = lm.my - x.get(1);
        double r = Math.sqrt(dx*dx + dy*dy);
        double b = wrapAngle(Math.atan2(dy, dx) - x.get(2));
        SimpleMatrix z = new SimpleMatrix(2,1);
        z.set(0, r);
        z.set(1, b);
        return z;
    }

    static SimpleMatrix jacobianH(SimpleMatrix x, Landmark lm) {
        double dx = lm.mx - x.get(0);
        double dy = lm.my - x.get(1);
        double q  = dx*dx + dy*dy;
        double eps = 1e-12;
        if (q < eps) q = eps;
        double r = Math.sqrt(q);

        SimpleMatrix H = new SimpleMatrix(2,3);
        // range
        H.set(0,0, -dx / r);
        H.set(0,1, -dy / r);
        // bearing
        H.set(1,0,  dy / q);
        H.set(1,1, -dx / q);
        H.set(1,2, -1.0);
        return H;
    }

    static int nearestLandmarkIndex(SimpleMatrix x, Landmark[] lms) {
        int best = 0;
        double bestd = 1e100;
        for (int i=0;i<lms.length;i++) {
            double dx = lms[i].mx - x.get(0);
            double dy = lms[i].my - x.get(1);
            double d = Math.sqrt(dx*dx + dy*dy);
            if (d < bestd) { bestd = d; best = i; }
        }
        return best;
    }

    public static void main(String[] args) {
        Landmark[] lms = new Landmark[] {
                new Landmark(5.0, 0.0),
                new Landmark(5.0, 5.0),
                new Landmark(0.0, 5.0),
                new Landmark(-3.0, 2.0)
        };

        double dt = 0.1;
        int N = 250;

        // Assumed process noise (tuning)
        SimpleMatrix Q = new SimpleMatrix(2,2);
        Q.set(0,0, 0.08*0.08);
        Q.set(1,1, 0.05*0.05);

        // Measurement noise
        SimpleMatrix R = new SimpleMatrix(2,2);
        R.set(0,0, 0.15*0.15);
        R.set(1,1, Math.pow(2.0*Math.PI/180.0, 2));

        // True noise (on control)
        double sigmaVTrue = 0.05;
        double sigmaWTrue = 0.03;

        Random rng = new Random(7);

        // Initial truth and estimate
        SimpleMatrix xTrue = new SimpleMatrix(3,1);
        SimpleMatrix xEst  = new SimpleMatrix(3,1);
        xEst.set(0, -0.2); xEst.set(1, 0.1); xEst.set(2, 0.05);

        SimpleMatrix P = new SimpleMatrix(3,3);
        P.set(0,0, 0.5*0.5);
        P.set(1,1, 0.5*0.5);
        P.set(2,2, Math.pow(10.0*Math.PI/180.0, 2));

        double sumPos2 = 0.0, sumTh2 = 0.0;

        for (int k=0;k<N;k++) {
            double vCmd = 0.8 + 0.2 * Math.sin(0.04 * k);
            double wCmd = 0.25 + 0.05 * Math.cos(0.03 * k);

            SimpleMatrix uCmd = new SimpleMatrix(2,1);
            uCmd.set(0, vCmd); uCmd.set(1, wCmd);

            // Truth uses noisy control
            double vTrue = vCmd + rng.nextGaussian() * sigmaVTrue;
            double wTrue = wCmd + rng.nextGaussian() * sigmaWTrue;
            SimpleMatrix uTrue = new SimpleMatrix(2,1);
            uTrue.set(0, vTrue); uTrue.set(1, wTrue);

            xTrue = motionModel(xTrue, uTrue, dt);

            // EKF predict
            SimpleMatrix F = jacobianF(xEst, uCmd, dt);
            SimpleMatrix L = jacobianL(xEst, dt);
            xEst = motionModel(xEst, uCmd, dt);
            P = F.mult(P).mult(F.transpose()).plus(L.mult(Q).mult(L.transpose()));
            P = P.plus(P.transpose()).scale(0.5);

            // Measurement (nearest landmark)
            int idx = nearestLandmarkIndex(xTrue, lms);
            Landmark lm = lms[idx];

            SimpleMatrix zTrue = measRangeBearing(xTrue, lm);
            double zR = zTrue.get(0) + rng.nextGaussian() * Math.sqrt(R.get(0,0));
            double zB = wrapAngle(zTrue.get(1) + rng.nextGaussian() * Math.sqrt(R.get(1,1)));
            SimpleMatrix zMeas = new SimpleMatrix(2,1);
            zMeas.set(0, zR); zMeas.set(1, zB);

            // EKF update
            SimpleMatrix H = jacobianH(xEst, lm);
            SimpleMatrix zHat = measRangeBearing(xEst, lm);
            SimpleMatrix y = zMeas.minus(zHat);
            y.set(1, wrapAngle(y.get(1)));

            SimpleMatrix S = H.mult(P).mult(H.transpose()).plus(R);
            SimpleMatrix K = P.mult(H.transpose()).mult(S.invert());

            xEst = xEst.plus(K.mult(y));
            xEst.set(2, wrapAngle(xEst.get(2)));

            SimpleMatrix I = SimpleMatrix.identity(3);
            P = (I.minus(K.mult(H))).mult(P).mult(I.minus(K.mult(H)).transpose())
                    .plus(K.mult(R).mult(K.transpose()));
            P = P.plus(P.transpose()).scale(0.5);

            // Errors
            double ex = xEst.get(0) - xTrue.get(0);
            double ey = xEst.get(1) - xTrue.get(1);
            double eth = wrapAngle(xEst.get(2) - xTrue.get(2));
            sumPos2 += ex*ex + ey*ey;
            sumTh2 += eth*eth;
        }

        double rmsPos = Math.sqrt(sumPos2 / (double)N);
        double rmsTh  = Math.sqrt(sumTh2 / (double)N);

        System.out.println("RMS position error [m]: " + rmsPos);
        System.out.println("RMS heading error [rad]: " + rmsTh);
    }
}
